File: Scala/src/com/yourtion/Pattern01/PersonExampleEx.scala
package com.yourtion.Pattern01

/**
  * Created by Yourtion on 9/7/16.
  */
object PersonExampleEx {
  case class Person(firstName: String, middleName: String, lastName: String)
  val p1 = Person("Aaron", "Jeffrey", "Smith")
  val p2 = Person("Aaron", "Bailey", "Zanthar")
  val p3 = Person("Brian", "Adams", "Smith")
  val people = Vector(p3, p2, p1)

  def complicatedSort(p1: Person, p2: Person) =
    if (p1.firstName != p2.firstName)
      p1.firstName < p2.firstName
    else if (p1.lastName != p2.lastName)
      p1.lastName < p2.lastName
    else
      p1.middleName < p2.middleName

  def run() = {
    people.sortWith(complicatedSort).foreach(p => println(p.firstName, p.middleName, p.lastName))
  }
}

File: Java/src/com/yourtion/Pattern01/PersonExampleEx.java
package com.yourtion.Pattern01;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;

/**
 * Created by Yourtion on 9/7/16.
 */
public class PersonExampleEx {
        public void run() {
            List<PersonExpanded> people = new ArrayList<PersonExpanded>();
            people.add(new PersonExpanded("Aaron", "Jeffrey", "Smith"));
            people.add(new PersonExpanded("Aaron", "Bailey", "Zanthar"));
            people.add(new PersonExpanded("Brian", "Adams", "Smith"));

            Collections.sort(people, new ComplicatedNameComparator());

            for (PersonExpanded person : people) {
                System.out.println(person);
            }
        }
}

class ComplicatedNameComparator implements Comparator<PersonExpanded> {
    public int compare(PersonExpanded p1, PersonExpanded p2) {

        int firstNameCompare =
                p1.getFirstName().compareTo(p2.getFirstName());
        int lastNameCompare = p1.getLastName().compareTo(p2.getLastName());
        int middleNameCompare = p1.getMidName().compareTo(p2.getMidName());

        if (0 != firstNameCompare)
            return firstNameCompare;
        else if (0 != lastNameCompare)
            return lastNameCompare;
        else
            return middleNameCompare;
    }
}